### PR TITLE
Vector store setup

### PR DESCRIPTION
**Pull request by diana-macrometa**
This PR explains how to start setting up Macrometa for a vector store. Specifically, it explains what to do first, how to create a collection, and different ways to get vector data into the collection.

### CODE REVIEW

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+Building on the setup process, let's explore how to effectively integrate semantic search capabilities with document collections in Macrometa, transforming it into a powerful vector store for complex data retrieval.
+
+---
+
+## Integrating Semantic Search with Document Collections
+
+Leveraging Macrometa for semantic search involves storing document data as vectors and utilizing Macrometa's querying capabilities to perform similarity searches. This section guides you through the preparation of document collections, indexing, and querying to achieve efficient semantic searches.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md]: "storing vectors as document data" seems more accurate

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+Semantic search transcends traditional keyword-based searches by understanding the context and meaning behind search queries. In the realm of vector stores, this involves converting text data into vectors that represent semantic meanings and using similarity searches to find the most relevant documents.
+
+### Preparing Document Collections for Vector Storage
+
+**Vectorization of Documents:**
+- The first step in preparing your document collection is to convert your documents into vectors. This typically involves using natural language processing (NLP) models, such as BERT or Word2Vec, to generate vector representations of text.
+- Each document is transformed into a high-dimensional vector that captures its semantic content, allowing for searches based on meaning rather than exact word matches.
+
+**Storing Vectors in Macrometa:**
+- Once your documents are vectorized, store these vectors in a Macrometa collection. Each vector becomes a document in the collection, with the vector dimensions stored as fields within the document.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md:20]: This is not correct. We expect one array in the document per vector dimension.

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+- Accompanying metadata, such as document titles, authors, or publication dates, can also be stored alongside the vectors to facilitate more complex queries and filtering.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md:22]: This is slightly misleading. This is the actual data that can be stored in addition to the vectors. The data is what is used to generate the vectors. Of course other data can be stored in the same document as well.

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+### Indexing Documents as Vectors
+
+To optimize the performance of semantic searches, it's crucial to index your document vectors effectively. Macrometa allows for the creation of custom indexes that can significantly speed up query times for specific types of searches.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md]: This is somewhat misleading too. For normal documents indexes merely speed up the search (documents can be searched without an index as well); however for vectors the index is essential.

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+- **Creating Indexes:** Depending on your specific use case, consider creating full-text indexes on metadata fields or geo-spatial indexes if your vectors represent spatial data. While direct indexing of high-dimensional vectors for similarity search is a complex challenge, these indexes can improve performance for searches that combine vector similarity with metadata filtering.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md]: why do we talk about fulltext or geo indexes here? Also what's the complex challenge here?

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+### Querying the Vector Store for Semantic Searches
+
+With your vectors stored and indexed in Macrometa, you're now ready to perform semantic searches.
+
+- **Writing Queries:** Use C8QL to write queries that find documents based on vector similarity. Although direct vector similarity searches require custom logic (for example, calculating cosine similarity through query functions), you can efficiently filter results based on metadata attributes using standard query syntax.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md]: why do we say it is "custom logic"? it seems to imply users having to write some code outside c8ql.

File: docs/vector-store/_4-integrate-semantic-search.md [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+- **Combining Filters:** For enhanced search capabilities, combine vector similarity calculations with metadata filters. This approach allows you to narrow down search results to the most relevant documents based on both their semantic content and metadata criteria.
```
Comment (amolth-macrometa) [inline docs/vector-store/_4-integrate-semantic-search.md]: as mentioned above not sure if we can call it "metadata"

File: docs/vector-store/_5-query-vector-store.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+```sql
+FOR doc IN documentCollection
+  FILTER doc.creationDate >= '2021-01-01' AND doc.creationDate <= '2021-12-31'
+  RETURN doc
+```
+This query returns documents created in the year 2021 from the collection `documentCollection`.
+
+### Performing Similarity Searches
+
+While Macrometa does not natively execute vector similarity searches directly through C8QL, you can implement functionality to perform such searches by integrating external vector search algorithms or by approximating similarity through query logic.
```
Comment (amolth-macrometa) [inline docs/vector-store/_5-query-vector-store.md]: why do you say we do not natively execute similarity searches? The entire sentence seems wrong to me.

File: docs/vector-store/_5-query-vector-store.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+- **Calculating Similarity:** Implement custom logic to calculate similarity scores between query vectors and document vectors stored in your collection. This might involve external processing to compute similarity metrics like cosine similarity or Euclidean distance and then querying Macrometa for documents closest to those scores.
```
Comment (amolth-macrometa) [inline docs/vector-store/_5-query-vector-store.md]: what is external processing?

File: docs/vector-store/_5-query-vector-store.md [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+
+To enhance the performance and relevance of your queries, consider the following advanced techniques:
+
+- **Use of Indexes:** Ensure that your queries leverage indexes effectively. For instance, if you have metadata that you frequently query alongside vector data, create and utilize indexes on those metadata fields to speed up search times.
+- **Combining Vector and Metadata Searches:** For more sophisticated search capabilities, combine your vector similarity logic with metadata-based filtering in your queries. This approach allows you to narrow down search results to documents that are not only similar in content but also meet specific metadata criteria.
+
+**Example of an Advanced Query:**
+```sql
+FOR doc IN documentCollection
+  FILTER doc.category == 'Technology' AND custom_similarity_function(doc.vector, @queryVector) > 0.9
```
Comment (amolth-macrometa) [inline docs/vector-store/_5-query-vector-store.md]: what is this custom function?

File: docs/vector-store/_6-vector-store-use-cases.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+- Apply clustering algorithms to categorize documents into clusters based on vector similarity.
+- For classification, train a model to identify the category of a document based on its vector, and use this model to automatically classify new documents.
+
+### Anomaly Detection in Data Streams
+
+**Overview:**
+- Anomaly detection identifies unusual patterns in data that do not conform to expected behavior. It's crucial for applications like fraud detection, system health monitoring, and identifying outliers in datasets.
+
+**Implementation:**
+- Stream data into Macrometa, converting each data point into a vector based on its features.
```
Comment (amolth-macrometa) [inline docs/vector-store/_6-vector-store-use-cases.md:35]: we don't have direct vector support for streams; user needs to store data in collection first

File: docs/vector-store/_7-vector-store-best-practices.md [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+### Efficient Indexing Strategies
+
+- **Index Design:** Design your indexes based on the queries you anticipate running most frequently. While direct indexing of high-dimensional vectors for similarity search is complex, creating indexes on associated metadata can significantly improve query performance.
+- **Selective Indexing:** Avoid over-indexing by only creating indexes that serve a specific purpose. Each additional index can add overhead to data insertion and update operations.
+
+### Scalability and Performance Optimization
+
+- **Monitor Query Performance:** Regularly monitor the performance of your queries using Macrometa’s analytics tools. Identify queries that are slow or resource-intensive and optimize them by refining the query structure or adjusting indexes.
+- **Data Partitioning:** Consider partitioning your data to improve query performance and manageability. Partitioning involves dividing your data into smaller, more manageable subsets, which can be queried more efficiently.
```
Comment (amolth-macrometa) [inline docs/vector-store/_7-vector-store-best-practices.md]: what exactly does this mean for us? we don't support partitioning directly.

File: docs/vector-store/_8-monitor-maintain-vector-store.md [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+- **Error Logs:** Regularly review error logs to catch and diagnose issues early. Macrometa provides logging capabilities that can alert you to problems affecting data integrity or query performance.
+- **Custom Analytics:** For deeper insights, consider integrating custom analytics tools or scripts that can analyze query patterns, usage trends, and operational anomalies. This can help in understanding user behavior and adjusting your vector store configuration accordingly.
+
+### Routine Maintenance Tasks and Checks
+
+- **Data Integrity Checks:** Periodally verify the integrity of your data. This includes checking for corruption, ensuring that backups are complete and restorable, and validating that data remains consistent across replicas.
+- **Index Optimization:** Review and optimize your indexes regularly. This might involve removing unused indexes, creating new indexes for frequently executed queries, or rebalancing indexes to improve performance.
```
Comment (amolth-macrometa) [inline docs/vector-store/_8-monitor-maintain-vector-store.md]: this probably does not apply to similarity searches as indexes are required in order to run the searches, so you cannot create new indexes for frequently executed queries (they have to pre-exist)

File: docs/vector-store/_9-troubleshoot-vector-store.md [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+
+## Troubleshooting Common Issues
+
+Even with meticulous setup and management, issues can emerge in any complex data system. Here are some common challenges encountered with vector stores like Macrometa, along with strategies for troubleshooting and resolution.
+
+### Identifying and Resolving Common Problems
+
+**Query Performance Degradation:**
+- **Symptoms:** Queries running slower than expected, timeouts, or partial results.
+- **Troubleshooting:** Check for inefficient queries that may not be using indexes effectively. Review the execution plan for your queries to identify bottlenecks.
```
Comment (amolth-macrometa) [inline docs/vector-store/_9-troubleshoot-vector-store.md]: doesn't apply to vector indexes

File: docs/vector-store/_9-troubleshoot-vector-store.md [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+- **Resolution:** Optimize query structures, adjust or create new indexes to improve performance, and ensure data is evenly distributed across shards if sharding is used.
```
Comment (amolth-macrometa) [inline docs/vector-store/_9-troubleshoot-vector-store.md]: distribution across shards is automatically managed by c8db.

File: docs/vector-store/_9-troubleshoot-vector-store.md [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+**Data Inconsistency:**
+- **Symptoms:** Mismatched query results, errors in data retrieval, or discrepancies in data replication.
+- **Troubleshooting:** Verify the integrity of your data. Look for replication errors or issues in the synchronization process between replicas.
+- **Resolution:** Re-sync affected data collections, repair corrupted data entries, and review replication configurations to prevent future inconsistencies.
+
+**Indexing Issues:**
```
Comment (amolth-macrometa) [inline docs/vector-store/_9-troubleshoot-vector-store.md]: doesn't apply

File: docs/vector-store/set-up-vector-store.md [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+### Familiarize Yourself with Macrometa
+
+Before you begin setting up your vector store in Macrometa, we recommend completing the [Quickstart](../quickstart.md) so that you are familiar with basic Macrometa tasks. The quickstart walks you through a simple exercise of creating and querying a collection.
+
+Also consider completing the following tutorials:
+
+- [Getting Started with Semantic Search](../search-views/semantic-search/getting-started-semantic-search.md)
+- [Query Tutorial](../queries/got-tutorial/)
+
+### Are Your Vectors Stored in an External Database?
```
Comment (amolth-macrometa) [inline docs/vector-store/set-up-vector-store.md:30]: this is not possible

File: docs/vector-store/set-up-vector-store.md [hunk 18]
```diff
@@ -0,0 +1,59 @@
...
+
+If so, then you might be able to use them as  an external source using Macrometa's Connections feature. Make sure they have a common schema and they are stored in a supported data source.
+
+A connection source _must_ be set up when you create the collection, so make sure everything in the source is prepared for import into Macrometa before you begin the process.
+
+For more information, refer to [Connections](../connections/).
+
+### Will Vectors be Streaming into the Collection?
```
Comment (amolth-macrometa) [inline docs/vector-store/set-up-vector-store.md:38]: this is not directly possible

File: docs/vector-store/set-up-vector-store.md [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+In order to better understand Macrometa [streams](../streams/) and [stream workers](../cep/), consider completing the following tutorials:
+
+- [Getting Started with Streams](../streams/getting-started-streams.md)
+- [Getting Started with Stream Workers](../cep/getting-started-stream-workers.md)
+
+## Create a Collection to Store Vectors
+
+In Macrometa, data is organized into collections, akin to tables in a relational database. To start using Macrometa as a vector store, you'll first need to create a new collection.
+
+- If you are using an external data source, then follow the instructions in [Create a Data Source Workflow](../connections/manage-etl-workflows.md#create-a-data-source-workflow). You can only perform this task if you have the Connections feature enabled.
```
Comment (amolth-macrometa) [inline docs/vector-store/set-up-vector-store.md:49]: not applicable

**Review by amolth-macrometa — CHANGES_REQUESTED** [review]

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 2 above)
Comment (diana-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md:20]: I haven't edited any of the files with the underscore yet. This PR is only for set-up-vector-store.md.

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 5 above)
Comment (diana-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md]: I'm going to rewrite this when I get to this topic in a future PR.

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 7 above)
Comment (diana-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md]: I think they're talking about metadata in the context of RAG here, which would be super simple to implement with GDN (theoretically).

File: docs/vector-store/_5-query-vector-store.md (hunk 8 above)
Comment (diana-macrometa) [reply docs/vector-store/_5-query-vector-store.md]: It probably is. I'll address it when I rewrite this topic in a future PR.

File: docs/vector-store/_5-query-vector-store.md (hunk 9 above)
Comment (diana-macrometa) [reply docs/vector-store/_5-query-vector-store.md]: No idea. Another thing I'll be rewriting in a future PR.

File: docs/vector-store/_5-query-vector-store.md (hunk 10 above)
Comment (diana-macrometa) [reply docs/vector-store/_5-query-vector-store.md]: Something I haven't tested and rewritten yet.

File: docs/vector-store/_6-vector-store-use-cases.md (hunk 11 above)
Comment (diana-macrometa) [reply docs/vector-store/_6-vector-store-use-cases.md:35]: Interesting! I'll dig into that a bit when I get to this topic.

File: docs/vector-store/set-up-vector-store.md (hunk 17 above)
Comment (diana-macrometa) [reply docs/vector-store/set-up-vector-store.md:30]: What I am referring to here is having the vectors in Snowflake or something and using a Connection to get them into the GDN. How is that not possible? Am I missing something?

File: docs/vector-store/set-up-vector-store.md (hunk 18 above)
Comment (diana-macrometa) [reply docs/vector-store/set-up-vector-store.md:38]: True, they would need to go via a stream worker, but in theory it is very possible.

File: docs/vector-store/set-up-vector-store.md (hunk 19 above)
Comment (diana-macrometa) [reply docs/vector-store/set-up-vector-store.md:49]: Why not? Isn't getting records from outside Macrometa into Macrometa the whole point of connection workflows?

**Comment by diana-macrometa** [comment]
Great review, @amolth-macrometa! For now, please just focus on set-up-vector-store.md. I haven't done any work on the other files, they will be addressed in future PRs.

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 2 above)
Comment (amolth-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md:20]: ok, would have been good to call it out in the description.

Comment (diana-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md:20]: I didn't think I'd made any changes to them, so didn't realize they had shown up. I will be more clear in the future.

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 7 above)
Comment (amolth-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md]: not sure I understand the relevance of RAG here

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 2 above)
Comment (diana-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md:20]: Oh, I get what I did wrong. I had the wrong base branch defined. Sorry!

File: docs/vector-store/_4-integrate-semantic-search.md (hunk 7 above)
Comment (diana-macrometa) [reply docs/vector-store/_4-integrate-semantic-search.md]: We'll hash it out in a future PR.

File: docs/vector-store/set-up-vector-store.md (hunk 17 above)
Comment (amolth-macrometa) [reply docs/vector-store/set-up-vector-store.md:30]: ok that is fine then

File: docs/vector-store/set-up-vector-store.md (hunk 19 above)
Comment (amolth-macrometa) [reply docs/vector-store/set-up-vector-store.md:49]: ok

File: docs/vector-store/set-up-vector-store.md [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+Before you begin setting up your vector store in Macrometa, we recommend completing the [Quickstart](../quickstart.md) so that you are familiar with basic Macrometa tasks. The quickstart walks you through a simple exercise of creating and querying a collection.
+
+Also consider completing the following tutorials:
+
+- [Getting Started with Semantic Search](../search-views/semantic-search/getting-started-semantic-search.md)
+- [Query Tutorial](../queries/got-tutorial/)
+
+### Are Your Vectors Stored in an External Database?
+
+If so, then you might be able to use them as  an external source using Macrometa's Connections feature. Make sure they have a common schema and they are stored in a supported data source.
```
Comment (amolth-macrometa) [inline docs/vector-store/set-up-vector-store.md]: extra space between "as" and "an"

**Review by amolth-macrometa — APPROVED** [review]

**Pull request merged**